Create TaskRecord model

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-18605

This is the data model that will be used to track record state, primarily to serve as a way to recover from a disaster scenario. This will have the benefit of tracking records that were not successfully sent to the broker in the first place.

~I'd like to keep these states as close to celery as possible, as it makes it easy to translate between the two as well, but am not sure what the most ideal way to do that is. I could have a test that just asserts that celery states are the strings I expect them to be to ensure data in this table doesn't accidentally diverge after a change to celery states that we don't notice.~ Since we will delete these records once the task has been processed, there isn't a need to have a field that tracks the exact state of the record, but rather just tracks the record while the task is in the pipeline to be processed.

~Open question: it feels wrong to have a "celery" app that I'm adding data models to, since in theory it could conflict with the celery package if they were to ever add models? Is it worth putting this elsewhere?~ Resolved below, but sticking with the `celery` app label.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Introduces a model that is not yet in use, so this change by itself is very safe.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
No tests.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->
Django will automatically delete this table if reversing.
 
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Create TaskRecord model

gherceg committed:
Ignore in data dumps and domain deletions

gherceg committed:
Ensure celery states do not diverge

millerdev commented on corehq/apps/celery/models.py:
```diff
@@ -0,0 +1,30 @@
+
+
+class TaskRecord(models.Model):
+    task_id = models.UUIDField(null=True)
```
Under what conditions does a task record have a null task id?

gherceg committed:
Merge branch 'master' into gh/task-record

gherceg committed:
Add RETRY and REJECTED states

gherceg committed:
Change app label to celery_tasks for distinctness

In the unlikely event that the celery python package adds a data model
in the future, this avoids any potential namespace conflicts, perhaps a
bit too prematurely.

gherceg committed:
Update deletion and dump model tests

millerdev commented on corehq/apps/celery/models.py:
```diff
@@ -0,0 +1,30 @@
+    STARTED = "STARTED", "Started"
+    FAILURE = "FAILURE", "Failure"
+    REVOKED = "REVOKED", "Revoked"
+    SUCCESS = "SUCCESS", "Success"
```
I can understand if you want to do this for testing, but I don't think we want them as a long-term solution. What do you think of adding these to a testing branch and roll that out on staging, but don't deploy them to prod?

millerdev commented on corehq/apps/celery/models.py:
```diff
@@ -4,12 +4,14 @@
+    FAILURE = "FAILURE", "Failure"
     PENDING = "PENDING", "Pending"  # task was created, but not yet sent to the broker
+    RECEIVED = "RECEIVED", "Received"  # task was received by a worker, not a broker
+    REJECTED = "REJECTED", "Rejected"  # not in docs, but set when task cannot be processed by worker
```
Is this a completed state? Should we just delete (or not create) the record if it's rejected?

gherceg committed:
Replace TaskRecordState with boolean

We really only need to track whether or not the task was sent to the
broker since the record is deleted once the task is processed.

gherceg committed:
Minor improvements to TaskRecord class

gherceg committed:
Add error field to TaskRecord

Optional field to capture error message if encountered when attempting
to create and send the task to the broker.

millerdev reviewed (approved):
Approved, assuming you have or will confirm that changing the app label didn't break anything.

gherceg committed:
Don't add custom app label

If celery every does introduce data models that conflict with this
namespace, they will likely need to have a widely available workaround,
so not going to worry about this/complicate our code in the meantime.

gherceg committed:
Mixed up auto_now and auto_now_add

auto_now_add only adds the datetime when the instance is first created,
so that is the one that we want for date_created.

gherceg committed:
Unique constraints in pg allow multiple null rows

https://www.postgresql.org/docs/current/ddl-constraints.html#DDL-CONSTRAINTS-UNIQUE-CONSTRAINTS

gherceg committed:
Forgot to update lock file after renaming migration

gherceg committed:
Remove last_modified field

Given the TaskRecord will be deleted upon completion, the only value the
last_modified field added was in the rare instance where a task was
retried with different args/kwargs, or upon retrying, ran into an error
queueing a message to the broker.

Given how rare either of these circumstances are, the existence of a
last_modified field is misleading in most cases, since a task could be
retried, but no changes are made, so last_modified wouldn't change. This
makes it a redundant field since created_at already holds this info.

gherceg committed:
Remove sent field

This field is redundant given the error field will indicate whether
there was an issue attempting to queue the task with more detail than a
boolean would give us anyway, and only needs to be set in the error
case which is more efficient.

gherceg committed:
Make task_id non-nullable and the primary key

Due to changes in the durable task implementation, every TaskRecord is
guaranteed to have a task_id.
